drivers/common/virtio_mi/lm.c

```diff
@@ -0,0 +1,1178 @@
+
+	rte_spinlock_unlock(&avq->lock);
+	return result->status;
+}
+
+int
+virtio_vdpa_cmd_identity(struct virtio_vdpa_pf_priv *priv,
+		struct virtio_admin_migration_identity_result *result)
+{
+	struct virtio_admin_migration_identity_result *res;
+	struct virtio_hw *hw = &priv->vpdev->hw;
+	struct virtio_admin_data_ctrl dat_ctrl;
+	struct virtio_admin_ctrl *ctrl;
+	int dlen[1];
+	int ret;
+
+	RTE_VERIFY(priv);
+
+	if (!virtio_with_feature(hw, VIRTIO_F_ADMIN_VQ)) {
+		CMD_LOG(INFO, "host does not support admin queue");
+		return -ENOTSUP;
+	}
+
+	ctrl = virtnet_get_aq_hdr_addr(hw->avq);
+	ctrl->hdr.class = VIRTIO_ADMIN_PCI_MIGRATION_CTRL;
+	ctrl->hdr.cmd = VIRTIO_ADMIN_PCI_MIGRATION_IDENTITY;
+	dlen[0] = 0;
+	dat_ctrl.num_in_data = 0;
+	dat_ctrl.num_out_data = 1;
+	dat_ctrl.out_data[0].iova = hw->avq->virtio_admin_hdr_mem
+				+ sizeof(struct virtio_admin_ctrl_hdr)
+				+ sizeof(ctrl->status);
+	dat_ctrl.out_data[0].len = sizeof(*result);
+
+	ret = virtio_vdpa_send_admin_command(hw->avq, ctrl, &dat_ctrl, dlen, 0);
+	if (ret) {
+		CMD_LOG(ERR, "Failed to run class %u, cmd %u, status %d", ctrl->hdr.class, ctrl->hdr.cmd, ret);
+		return ret;
+	}
+
+	res = (struct virtio_admin_migration_identity_result *)
+			rte_mem_iova2virt(dat_ctrl.out_data[0].iova);
+	result->major_ver = rte_le_to_cpu_16(res->major_ver);
+	result->minor_ver = rte_le_to_cpu_16(res->minor_ver);
+	result->ter_ver   = rte_le_to_cpu_16(res->ter_ver);
+
+	return ret;
+}
+
+static int
+virtio_vdpa_cmd_set_status(struct virtio_hw *hw, uint16_t vdev_id,
+		enum virtio_internal_status status)
+{
+	struct virtio_admin_migration_modify_internal_status_data *sd;
+	struct virtio_admin_data_ctrl dat_ctrl;
+	struct virtio_admin_ctrl *ctrl;
+	int dlen[1];
+	int ret;
+
+	ctrl = virtnet_get_aq_hdr_addr(hw->avq);
+	ctrl->hdr.class = VIRTIO_ADMIN_PCI_MIGRATION_CTRL;
+	ctrl->hdr.cmd = VIRTIO_ADMIN_PCI_MIGRATION_MODIFY_INTERNAL_STATUS;
+	sd = (struct virtio_admin_migration_modify_internal_status_data *)&ctrl->data[0];
+	sd->vdev_id = rte_cpu_to_le_16(vdev_id);
+	sd->internal_status = rte_cpu_to_le_16((uint16_t)status);
+	dlen[0] = sizeof(*sd);
+	dat_ctrl.num_in_data = 0;
+	dat_ctrl.num_out_data = 0;
+
+	ret = virtio_vdpa_send_admin_command(hw->avq, ctrl, &dat_ctrl, dlen, 1);
+	if (ret) {
+		CMD_LOG(ERR, "Failed to run class %u, cmd %u, status %d, vdev_id: %u",
+				ctrl->hdr.class, ctrl->hdr.cmd, ret, vdev_id);
+		return ret;
+	}
+
+	return 0;
+}
+
+int
+virtio_vdpa_cmd_resume(struct virtio_vdpa_pf_priv *priv, uint16_t vdev_id,
+		enum virtio_internal_status status)
+{
+	struct virtio_hw *hw;
+
+	RTE_VERIFY(priv);
+
+	if (status != VIRTIO_S_QUIESCED && status != VIRTIO_S_RUNNING)
+		return -EINVAL;
+
+	hw = &priv->vpdev->hw;
+	if (!virtio_with_feature(hw, VIRTIO_F_ADMIN_VQ)) {
+		CMD_LOG(INFO, "host does not support admin queue");
+		return -ENOTSUP;
+	}
+
+	return virtio_vdpa_cmd_set_status(hw, vdev_id, status);
+}
+
+int
+virtio_vdpa_cmd_suspend(struct virtio_vdpa_pf_priv *priv, uint16_t vdev_id,
```

Comment:
we should also have get status API
VIRTIO_ADMIN_PCI_MIGRATION_GET_INTERNAL_STATUS